src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+                    { headerOffset = offset
+                    , headerPayloadOffset = offset + 512
+                    , headerFileNameSuffix = suffix
+                    , headerFileMode = fileMode fi
+                    , headerOwnerId = fileUserId fi
+                    , headerGroupId = fileGroupId fi
+                    , headerPayloadSize = payloadSize
+                    , headerTime = fileModTime fi
+                    , headerLinkIndicator = linkIndicator
+                    , headerLinkName = linkName
+                    , headerMagicVersion = ustarMagicVersion
+                    , headerOwnerName = toShort $ fileUserName fi
+                    , headerGroupName = toShort $ fileGroupName fi
+                    , headerDeviceMajor = 0
+                    , headerDeviceMinor = 0
+                    , headerFileNamePrefix = prefix
+                    }
+
+
+-- | Split a file path at the @n@ mark from the end, while still keeping the
+-- split as a valid path, i.e split at a path separator only.
+splitPathAt :: Int -> ByteString -> (ShortByteString, ShortByteString)
+splitPathAt n fp
+    | S.length fp <= n = (SS.empty, toShort fp)
+    | otherwise =
+        let sfp = S8.splitWith isPathSeparator fp
+            sepWith p (tlen, prefix', suffix') =
+                case S.length p + 1 + tlen of
+                    tlen'
+                        | tlen' <= n -> (tlen', prefix', p : suffix')
+                    tlen' -> (tlen', p : prefix', suffix')
+            (_, prefix, suffix) = foldr' sepWith (0, [], []) sfp
+            toShortPath = toShort . S8.intercalate pathSeparatorS
+        in (toShortPath prefix, toShortPath suffix)
+
+packHeader :: MonadThrow m => Header -> m S.ByteString
+packHeader header = do
+    (left, right) <- packHeaderNoChecksum header
+    let sumsl :: SL.ByteString -> Int
+        sumsl = SL.foldl' (\ !acc !v -> acc + fromIntegral v) 0
+    encChecksum <- encodeOctal 7 $ sumsl left + 32 * 8 + sumsl right
+    return $
+        SL.toStrict $
+        toLazyByteString $ lazyByteString left <> encChecksum <> word8 0 <> lazyByteString right
+
+
+packHeaderNoChecksum :: MonadThrow m => Header -> m (SL.ByteString, SL.ByteString)
+packHeaderNoChecksum Header {..} = do
+    let CTime headerTime' = headerTime
+    hNameSuffix <- encodeShort 100 headerFileNameSuffix
+    hFileMode <- encodeOctal 7 headerFileMode
+    hOwnerId <- encodeOctal 7 headerOwnerId
+    hGroupId <- encodeOctal 7 headerGroupId
+    hPayloadSize <- encodeOctal 11 headerPayloadSize
+    hTime <- encodeOctal 11 headerTime'
+    hLinkName <- encodeShort 100 headerLinkName
+    hMagicVersion <- encodeShort 8 headerMagicVersion
+    hOwnerName <- encodeShort 32 headerOwnerName
+    hGroupName <- encodeShort 32 headerGroupName
+    hDevMajor <- encodeDevice headerDeviceMajor
+    hDevMinor <- encodeDevice headerDeviceMinor
+    hNamePrefix <- encodeShort 155 headerFileNamePrefix
+    return
+        ( toLazyByteString $
```

Comment:
Would it make sense to just return the builder itself instead, and have the API focus on builders?

Reply:
You gotta convert to ByteString in order to compute the checksum, that's the only reason for doing it that way.

Reply:
Fair enough 👍